sopel/modules/remind.py

```diff
@@ -226,47 +226,227 @@ def remind_in(bot, trigger):
+    r'\s+(?P<message>.*)'
+)
+
+
+class TimeReminder(object):
+    """Time reminder for the ``at`` command"""
+    def __init__(self,
+                 hour,
+                 minute,
+                 second,
+                 timezone,
+                 date1,
+                 date2,
+                 date3,
+                 message):
+        self.hour = hour
+        self.minute = minute
+        self.second = second
+        self.timezone = pytz.timezone(timezone)
+        self.message = message
+
+        year = None
+        month = None
+        day = None
+
+        if date1 and date2 and date3:
+            if len(date1) == 4:
+                # YYYY-mm-dd
+                year = int(date1)
+                month = int(date2)
+                day = int(date3)
+            else:
+                # dd-mm-YYYY or dd/mm/YY
+                year = int(date3)
+                month = int(date2)
+                day = int(date1)
+        elif date1 and date2:
+            if len(date1) == 4:
+                # YYYY-mm
+                year = int(date1)
+                month = int(date2)
+            elif len(date2) == 4:
+                # mm-YYYY
+                year = int(date2)
+                month = int(date1)
+            else:
+                # dd/mm
+                month = int(date2)
+                day = int(date1)
+
+        self.year = year
+        self.month = month
+        self.day = day
+
+    def __eq__(self, other):
+        return all(
+            getattr(self, attr) == getattr(other, attr, None)
+            for attr in [
+                'hour',
+                'minute',
+                'second',
+                'timezone',
+                'year',
+                'month',
+                'day',
+                'message',
+            ]
+        )
+
+    def get_duration(self, today=None):
+        """Get the duration between the reminder and ``today``
+
+        :param today: aware datetime to compare to, default to UTC now
+        :type today: aware :class:`datetime.datetime`
+        :return: The duration, in second, between ``today`` and the reminder.
+        :rtype: :class:`int`
+
+        This method returns the number of seconds given by the
+        :class:`datetime.timedelta` obtained by the difference between the
+        reminder and ``today``.
+
+        If the delta between the reminder and ``today`` is negative, Python
+        will represent it as a negative number of days, and a positive number
+        of seconds: since it returns the number of seconds, any past reminder
+        will be transformed into a future reminder the next day.
+        """
+        if not today:
+            today = datetime.now(self.timezone)
+        else:
+            today = today.astimezone(self.timezone)
+
+        year = self.year if self.year is not None else today.year
+        month = self.month if self.month is not None else today.month
+        day = self.day if self.day is not None else today.day
```

Comment:
Is it not safe enough to do e.g. `year = self.year or today.year`? When might `self.year` be not `None` but still falsy?

Reply:
No, because I convert beforehand `date1` to an integer, so `self.year` could be `0`, and `0 or today.year` returns `today.year`. It means that this would not raise:

    <@Exirel> .at 13:37 0000-00-00 the first message between BC and AD

and instead it would consider today's date. If user sends an invalid input, we should gently scream at them.

Quick note from the [Zen of Python](https://www.python.org/dev/peps/pep-0020/):

> In the face of ambiguity, refuse the temptation to guess.